### PR TITLE
Annotate layout variables at the ends of layout errors

### PR DESCRIPTION
**Pull request by antalsz**
This improves how some layout errors are printed, so you no longer have the confusing `'a should be an instance of type 'b`.

This also does some refactoring of where things live for dependency and clarity reasons!

### CODE REVIEW

File: lambda/translcore.ml [hunk 1]
```diff
@@ -244,7 +245,7 @@ let rec push_defaults loc bindings use_lhs cases partial warnings =
               Texp_ident
                 (Path.Pident param, mknoloc (Longident.Lident name),
                  desc, Id_value)},
-             Types.Sort.value,
+             Layouts.Sort.value,
```
Comment (goldfirere) [inline lambda/translcore.ml]: `Layouts` is open, so this prefix is redundant.

File: lambda/translcore.ml [hunk 2]
```diff
@@ -1048,7 +1049,7 @@ and transl_exp0 ~in_new_scope ~scopes void_k e =
         let {val_type; _} = Env.find_value (Pident id) e.exp_env in
         match
           Ctype.check_type_layout e.exp_env (Ctype.correct_levels val_type)
-            Types.Layout.value
+            Layouts.Layout.value
```
Comment (goldfirere) [inline lambda/translcore.ml]: `Layouts` is open, so the prefix is redundant.

File: typing/oprint.ml [hunk 3]
```diff
@@ -277,6 +277,10 @@ let join_modes rm1 am2 =
   | _, Oam_unknown -> Oam_unknown
   | Oam_global, Oam_global -> Oam_global
 
+let print_out_layout ppf = function
+  | Olay_const lay -> fprintf ppf "%s" (Layouts.Layout.string_of_const lay)
+  | Olay_var       -> fprintf ppf "'_concrete_layout"
```
Comment (goldfirere) [inline typing/oprint.ml]: Do we actually ever print this? If we don't, then we should remove this code. If we do, then we'll have to be careful to e.g. number the `concrete_layout` variables to make them not all look like aliases. That's a small bit of work and needn't be in this commit. But this line of code makes me a little nervous.

File: typing/outcometree.mli [hunk 4]
```diff
@@ -69,6 +69,10 @@ type out_global =
   | Ogf_nonlocal
   | Ogf_unrestricted
 
+type out_layout =
+  | Olay_const of Asttypes.const_layout
+  | Olay_var (* XXX ASZ do we need to invent names for these? *)
```
Comment (goldfirere) [inline typing/outcometree.mli]: As I say in my other comment: yes, sadly.

File: typing/outcometree.mli [hunk 5]
```diff
@@ -88,6 +92,9 @@ type out_type =
   | Otyp_poly of string list * out_type
   | Otyp_module of out_ident * (string * out_type) list
   | Otyp_attribute of out_type * out_attribute
+  | Otyp_layout_annot of out_type * out_layout
+      (* Currently only introduced with very explicit code in [Printtyp] and not
```
Comment (goldfirere) [inline typing/outcometree.mli:96]: :heart:

**Review by goldfirere — COMMENTED** [review]
Love the new error messages!

File: typing/oprint.ml (hunk 3 above)
Comment (antalsz) [reply typing/oprint.ml]: I don't think we currently do, so are we good to leave this handled by the `XXX ASZ` comments for now?  I'll add one here, too

**Comment by antalsz** [comment]
If we're comfortable leaving the `Olay_var` stuff under `XXX`s, then we're ready to merge; @ccasin, how should I get that right since the code moved?

**Comment by ccasin** [comment]
@antalsz I think just open a new PR against that branch. The code hasn't changed much so it should be an easy rebase.

File: typing/oprint.ml (hunk 3 above)
Comment (goldfirere) [reply typing/oprint.ml]: I'd still rather have something more loudly broken here: either an outright `failwith` or a `"if you see this, we have a bug"`. And I believe our convention is to have `CR layouts` as the search string for issues to be merged but then to return to. `XXX` is something we won't merge with `main`.

Comment (antalsz) [reply typing/oprint.ml]: I changed the variable name to `'_concrete_layout__should_never_be_printed` and updated the two comments to `CR layouts`

**Review by goldfirere — APPROVED** [review]
Ready to merge. Except that it shouldn't be, because we're now on the other branch.

**Comment by ccasin** [comment]
Great.  @antalsz if you close this PR and open a new one over there we can just merge it.  The only extra commit on that branch is me addressing some review, so this shouldn't cause serious conflicts or need re-review.

**Comment by antalsz** [comment]
Moved to ocaml-flambda#147

**Pull request closed without merging**